src/index.ts

```diff
@@ -1,3 +1,182 @@
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
 'use-strict';
+
+import {
+  IInputAreaViewModel
+} from 'jupyter-js-input-area';
+
+import {
+  IOutputAreaViewModel
+} from 'jupyter-js-output-area';
+
+import {
+  IObservableList
+} from 'phosphor-observablelist';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+import './index.css';
+
+
+/**
+ * An enum which describes the type of cell.
+ */
+enum CellType {
+  /**
+   * The cell contains code input.
+   */
+  Code,
+
+  /**
+   * The cell contains markdown.
+   */
+  Markdown,
+
+  /**
+   * The cell contains raw text.
+   */
+  Raw
+}
+
+
+/**
+ * An enum which describes which interface property has changed.
+ */
+export
+enum CellProperty {
+  CellType,
+  Tags,
+  Deletable,
+  Mergable,
+  Splittable,
+  Rendered,
+  Marked,
+  Format
+}
+
+
+/**
+ * The arguments object emitted with the `stateChanged` signal.
+ */
+export
+interface ICellChangedArgs<T> {
+  which: CellProperty,
+  oldValue: T;
+  newValue: T;
+}
+
+
+/**
+ * The definition of a model object for a base cell.
+ */
+interface IBaseCellViewModel {
+
+  /**
+   * The type of cell.
+   */
+  cellType: CellType;
+
+  /**
+   * Tags applied to the cell.
+   */
+  tags?: IObservableList<string>;
+
+  /**
+   * Get namespaced metadata about the cell.
+   */
+  getMetadata(namespace: string) : IObservableMap<string, JSONData>;
```

Comment:
In principle each library that starts to use a particular namespace in the
metadata could formally type the objects that appear in their namespace.Two
examples that we have today:

* Slide show. Various parts of our code base can present a slide show view
of the notebook. The cell metadata is used to encode how each cell with
function in that slide show. This metadata is edited by a UI provided by a
cell toolbar UI.
* nbgrader. nbgrader has a cell toolbar UI that allows an instructor to
encode how a cell will function in an assignment (test, points, etc.). Here
are the docs for that:

http://nbgrader.readthedocs.org/en/stable/user_guide/02_developing_assignments.html

* You could also imagine that the dashboard layout could get written to the
cell/notebook metadata so when I email you a notebook, you can immediately
access my dashboard layout.

All of this "data" is truly "metadata" in that it looses its meaning unless
it is linked to the cell/notebook.

There are many other libraries that are using notebook/cell metadata in
other ways.

The problem is that all of the cell toolbar UIs will be provided by
separate npm packages/plugins that provide celltoolbarview extensions to
some extension point. At the time we compile the code for notebook/cells,
we have no idea what the structural form of that metadata will be at
runtime. If we could know that in advance, then yes, we could formally type
it all.

However, it very well may be that the celltoolbar extension authors will
actually formally type this data in their own code - they will know that
the data looks like at compile time when types matter. Thus as long as they
only read/write the metadata object in their own namespace, the compiler
can promise type safety. This *would* help these plugins formalize the
information being stored in these fields.

On Tue, Nov 3, 2015 at 8:27 PM, S. Chris Colbert <notifications@github.com>
wrote:

> In src/index.ts
> <https://github.com/jupyter/jupyter-js-cells/pull/2#discussion_r43832237>:
>
> > +interface IBaseCellViewModel {
> > +
> > +  /**
> > +   * The type of cell.
> > +   */
> > +  cellType: CellType;
> > +
> > +  /**
> > +   * Tags applied to the cell.
> > +   */
> > +  tags?: IObservableList<string>;
> > +
> > +  /**
> > +   * Get namespaced metadata about the cell.
> > +   */
> > +  getMetadata(namespace: string) : IObservableMap<string, JSONData>;
>
> I'm not clear *what* will be edited by these toolbars, and why that data
> cant be formally typed, or at least exposed at a level below these view
> models. I think a concrete use case would help me understand here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/jupyter-js-cells/pull/2/files#r43832237>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com